### DISCUSSION TITLE
Is passing the notifier as state safe?

### DISCUSSION THREAD

**Discussion by thoggs** [rrousselGit/riverpod#2914] [General]
Hello everybody,

I would like to know if passing the Notifier as State to the AsyncDisposedNotifier class is safe and recommended and what problems can we have with this approach?

![IMG_0362](https://github.com/rrousselGit/riverpod/assets/43221193/de6ec52b-cb74-4d98-8ca8-e8cb039817f8)

**Comment by AhmedLSayed9** [comment]
That's fine.
However, I'd prefer using basic `Provider` that just return a class instance of `HomeViewModel`.

**Reply by thoggs** [reply]
It would be an alternative, but it would change the architectural scope we are planning for the application.

**Reply by AhmedLSayed9** [reply]
I meant [Provider](https://pub.dev/documentation/riverpod/latest/riverpod/Provider-class.html) from Riverpod

**Comment by snapsl** [comment]
@thoggs By setting the notifier as state u can access the notifier properties and methods like `ref` and `build()` from the `state`. This seems weird. It will probably not break anything, but I wouldn't call it recommended.

**Reply by thoggs** [reply]
I understood!

When we talk about architecture, it really doesn't seem very healthy to me, especially if the project grows a lot.

In this case, the safest option would be to create a separate state class, as per the RiverPod documentation?

**Reply by snapsl** [reply]
Yes, a separate state class sounds good.
Long term, also have a look into code generation where the notifier type is selected automatically for you.
